packages/gravity/kube-testing/src/analysys/stat-analysys.ts

```diff
@@ -0,0 +1,165 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { Series } from 'danfojs-node';
+
+import { PlanResults } from '../plan/spec-base';
+import { LogReader, TraceEvent } from './reducer';
+
+const seriesToJson = (s: Series) => {
+  const indexes = s.index;
+  return Object.fromEntries(
+    (s.values as Array<number>).map((val, i) => {
+      return [indexes[i], val];
+    })
+  );
+};
+
+const getStats = (series: number[], additionalMetrics: Record<string, number> = {}) => {
+  const stats = new Series(series).describe() as Series;
+
+  const values: number[] = [];
+  const indexes: string[] = [];
+  Object.entries(additionalMetrics).forEach(([key, value]) => {
+    values.push(value);
+    indexes.push(key);
+  });
+
+  stats.append(values, indexes, { inplace: true });
+  stats.print();
+
+  return seriesToJson(stats);
+};
+
+export const mapToJson = (m: Map<string, any>) => {
+  return Object.fromEntries(
+    Array.from(m.entries()).map(([key, val]) => {
+      let decoded = val;
+      if (val instanceof Map) {
+        decoded = mapToJson(val);
+      }
+      return [key, decoded];
+    })
+  );
+};
+
+export const analyzeMessages = async (results: PlanResults) => {
+  const messages = new Map<string, { sent?: number; received?: number }>();
+  const reader = new LogReader();
+
+  for (const { logFile } of Object.values(results.agents)) {
+    reader.addFile(logFile);
+  }
+
+  for await (const entry of reader) {
+    if (entry.message !== 'dxos.test.signal') {
+      continue;
+    }
+    const data: TraceEvent = entry.context;
+
+    if (data.type !== 'SENT_MESSAGE' && data.type !== 'RECEIVE_MESSAGE') {
+      continue;
+    }
+    if (!messages.has(data.message)) {
+      messages.set(data.message, { sent: undefined, received: undefined });
+    }
+
+    switch (data.type) {
+      case 'SENT_MESSAGE':
+        messages.get(data.message)!.sent = entry.timestamp;
+        break;
+      case 'RECEIVE_MESSAGE':
+        messages.get(data.message)!.received = entry.timestamp;
+        break;
+    }
+  }
+
+  const failures = Array.from(messages.values()).filter((x) => !x.received || !x.sent).length;
+  const lagTimes = Array.from(messages.values())
+    .filter((x) => !!x.sent && !!x.received)
+    .map((x) => x.received! - x.sent!);
+  console.log('Succesfull messages', lagTimes.length);
+
+  return getStats(lagTimes, { failures });
+};
+
+export const analyzeSwarmEvents = async (results: PlanResults) => {
+  /**
+   * topic -> peerId -> { join: time, left: time, seen: peerId -> ts}
+   */
+  const topics = new Map<string, Map<string, { join?: number; leave?: number; seen: Map<string, number> }>>();
+
+  const reader = new LogReader();
+
+  for (const { logFile } of Object.values(results.agents)) {
+    reader.addFile(logFile);
+  }
+
+  for await (const entry of reader) {
```

Comment:
for long functions please consider adding a single line comment for these multiple loops.

// Reduce log events into stats keyed by topic.

BELOW: 

// Reduce topic stats into metrics (failures, latency)

And again if this were a class, it would be simpler/more compact to split these into methods, which would possibly be more stable over time (i.e., adding more methods).

```
const analyzer = new Analyzer(results);
analyzer.processErrors();
analyzer.processLatency();
analyzer.processSwarmEvents();
analyzer.somethingWeDoNext();
```

And define type for Stats that Analyzer accumulates?